File: src/Scanner.Core/Checks/MissingHeadingStructureCheck.cs
using System.Text.RegularExpressions;
using Scanner.Core.Rules;

namespace Scanner.Core.Checks;

/// <summary>
/// Checks for documents that do not include any heading elements.
/// </summary>
public sealed class MissingHeadingStructureCheck : ICheck
{
    private static readonly Regex HeadingRegex = new("<h[1-6][^>]*>", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    /// <inheritdoc />
    public string Id => "missing-heading-structure";

    /// <inheritdoc />
    public IReadOnlyCollection<string> ApplicableKinds { get; } = new[] { "html", "htm", "cshtml", "razor" };

    /// <inheritdoc />
    public IEnumerable<Issue> Run(CheckContext context, RuleDefinition rule)
    {
        if (HeadingRegex.IsMatch(context.Content))
        {
            return Array.Empty<Issue>();
        }

        var line = TextUtilities.GetLineNumber(context.Content, 0);
        return new[]
        {
            new Issue(rule.Id, Id, context.FilePath, line, "Document has no heading structure.", null)
        };
    }
}

File: node_engine/src/checks/MissingHeadingStructureCheck.js
const { getLineNumber } = require("./TextUtilities");

const headingRegex = /<h[1-6][^>]*>/i;

const MissingHeadingStructureCheck = {
  id: "missing-heading-structure",
  applicableKinds: ["html", "htm", "cshtml", "razor"],
  run(context, rule) {
    if (context.document?.querySelector) {
      if (context.document.querySelector("h1, h2, h3, h4, h5, h6")) {
        return [];
      }

      return [
        {
          ruleId: rule.id,
          checkId: MissingHeadingStructureCheck.id,
          filePath: context.filePath,
          line: getLineNumber(context.content, 0),
          message: "Document has no heading structure.",
          evidence: null
        }
      ];
    }

    if (headingRegex.test(context.content)) {
      return [];
    }

    return [
      {
        ruleId: rule.id,
        checkId: MissingHeadingStructureCheck.id,
        filePath: context.filePath,
        line: getLineNumber(context.content, 0),
        message: "Document has no heading structure.",
        evidence: null
      }
    ];
  }
};

module.exports = { MissingHeadingStructureCheck };
